wen-restart/src/heaviest_fork_aggregate.rs

```diff
@@ -0,0 +1,560 @@
+            (my_heaviest_fork_slot, my_heaviest_fork_hash),
+            Self::validator_stake(epoch_stakes, my_pubkey),
+        );
+        Self {
+            supermajority_threshold: wait_for_supermajority_threshold_percent as f64 / 100.0,
+            my_shred_version,
+            my_pubkey: *my_pubkey,
+            epoch_stakes: epoch_stakes.clone(),
+            heaviest_forks: HashMap::new(),
+            block_stake_map,
+            active_peers,
+            active_peers_seen_supermajority: HashSet::new(),
+        }
+    }
+
+    // TODO(wen): this will a function in separate EpochStakesMap class later.
+    fn validator_stake(epoch_stakes: &EpochStakes, pubkey: &Pubkey) -> u64 {
+        epoch_stakes
+            .node_id_to_vote_accounts()
+            .get(pubkey)
+            .map(|x| x.total_stake)
+            .unwrap_or_default()
+    }
+
+    pub(crate) fn aggregate_from_record(
+        &mut self,
+        key_string: &str,
+        record: &HeaviestForkRecord,
+    ) -> Result<Option<HeaviestForkRecord>> {
+        let from = Pubkey::from_str(key_string)?;
+        let bankhash = Hash::from_str(&record.bankhash)?;
+        let converted_record = RestartHeaviestFork {
+            from,
+            wallclock: record.wallclock,
+            last_slot: record.slot,
+            last_slot_hash: bankhash,
+            observed_stake: record.total_active_stake,
+            shred_version: record.shred_version as u16,
+        };
+        Ok(self.aggregate(converted_record))
+    }
+
+    pub(crate) fn aggregate(
+        &mut self,
+        received_heaviest_fork: RestartHeaviestFork,
+    ) -> Option<HeaviestForkRecord> {
+        let total_stake = self.epoch_stakes.total_stake();
+        let from = &received_heaviest_fork.from;
+        let sender_stake = Self::validator_stake(&self.epoch_stakes, from);
+        if sender_stake == 0 {
+            warn!(
+                "Gossip should not accept zero-stake RestartLastVotedFork from {:?}",
+                from
+            );
+            return None;
+        }
+        if received_heaviest_fork.shred_version != self.my_shred_version {
+            warn!(
+                "Gossip should not accept RestartLastVotedFork with different shred version {} from {:?}",
+                received_heaviest_fork.shred_version, from
+            );
+            return None;
+        }
+        let record = HeaviestForkRecord {
+            slot: received_heaviest_fork.last_slot,
+            bankhash: received_heaviest_fork.last_slot_hash.to_string(),
+            total_active_stake: received_heaviest_fork.observed_stake,
+            shred_version: received_heaviest_fork.shred_version as u32,
+            wallclock: received_heaviest_fork.wallclock,
+        };
+        if let Some(old_heaviest_fork) = self
+            .heaviest_forks
+            .insert(*from, received_heaviest_fork.clone())
+        {
+            if old_heaviest_fork == received_heaviest_fork {
```

Comment:
I think we should directly compare slot and hash here so that other factors aren't in play like wallclock

Reply:
Done